### PR TITLE
network policies for user cluster kube-system namespace

### PR DESCRIPTION
**Pull request by mate4st**
**What this PR does / why we need it**:

Introduce network policies for user cluster in kube-system namespace

Fixes https://github.com/kubermatic/edgematic/issues/63

Done with: https://networkpolicy.io/

Feature can be disabled via feature flag **kubeSystemNetworkPolicies** at cluster spec. It is automatically added to new clusters, existing cluster won't be affected.

Attention: 
In case any custom applications are deployed to the user clusters kube-system namespace that need specific network access, a network policy needs to be added for that application to allow traffic other than dns.

```release-note
Deploy network policies with default deny to kube-system namespace in user cluster. Adding network policies for custom deployments in kube-system namespace maybe required.
```

### CODE REVIEW

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+	"k8c.io/kubermatic/v2/pkg/controller/operator/common"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+)
+
+// AllowAllDnsNetworkPolicyCreator allows egress traffic from all pods to kube-dns
+func AllowAllDnsNetworkPolicyCreator() reconciling.NamedNetworkPolicyCreatorGetter {
+	return func() (string, reconciling.NetworkPolicyCreator) {
+		return "allow-dns", func(np *networkingv1.NetworkPolicy) (*networkingv1.NetworkPolicy, error) {
+			dnsPort := intstr.FromInt(53)
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: nit: Could be just `fmt.Sprint`.

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+							},
+							{
+								Protocol: &protoUdp,
+								Port:     &dnsPort,
+							},
+						},
+						To: []networkingv1.NetworkPolicyPeer{
+							{
+								IPBlock: &networkingv1.IPBlock{
+									CIDR: "169.254.20.10/32",
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: `fmt.Sprintf("%s/32", somepackge.NodeLocalDNSCacheAddress)`

File: addons/csi/hetzner/hcloud-csi-1.6.0.yaml [hunk 3]
```diff
@@ -364,6 +364,22 @@ spec:
...
+metadata:
+  name: hcloud-csi
+  namespace: kube-system
+spec:
+  podSelector:
+    matchLabels:
+      app: hcloud-csi
+  policyTypes:
+    - Egress
+  egress:
```
Comment (PratikDeoghare) [inline addons/csi/hetzner/hcloud-csi-1.6.0.yaml:390]: Isn't this same as `egress: {}`?

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"k8c.io/kubermatic/v2/pkg/controller/operator/common"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+)
+
+// AllowAllDnsNetworkPolicyCreator allows egress traffic from all pods to kube-dns
+func AllowAllDnsNetworkPolicyCreator() reconciling.NamedNetworkPolicyCreatorGetter {
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: nit: `Dns -> DNS`

**Comment by rastislavs** [comment]
A general question to this PR: are we really confident enough that this won't break anything to roll this out unconditionally to all existing user clusters of all existing KKP installations which upgrade to the new KKP version?

I would choose a bit safer approach, similar to the one chosen for apiserver network policies:

- Have a cluster API option (e.g. in `ClusterNetworkingConfig`) / cluster feature gate (like this one for [apiserver policies](https://github.com/kubermatic/kubermatic/blob/18c98aefc7ad2c9713e3bb34cd62f7e61d17ae5a/pkg/crd/kubermatic/v1/cluster.go#L231-L233)) for enabling these network policies,
- If these policies have to be enabled by default for each user cluster, then do it only for newly created clusters, e.g. in the mutation webhook like for [apiserver network policies](https://github.com/kubermatic/kubermatic/blob/18c98aefc7ad2c9713e3bb34cd62f7e61d17ae5a/pkg/webhook/cluster/mutation/mutation.go#L165-L171),
- In any case, it would be a good idea to have an option to disable these network polices - for special cases, if something breaks in production environment or for debugging purposes.

File: addons/csi/hetzner/hcloud-csi-1.6.0.yaml (hunk 3 above)
Comment (rastislavs) [reply addons/csi/hetzner/hcloud-csi-1.6.0.yaml:390]: same question :) If they are same, I would prefer to use the same expression everywhere, ideally the one which will be compatible with IPv6 ;)

File: addons/openvpn/networkpolicy.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# Copyright 2021 The Kubermatic Kubernetes Platform contributors.
```
Comment (rastislavs) [inline addons/openvpn/networkpolicy.yaml:1]: we would need a similar policy for konnectivity-agent as well

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+	return func() (string, reconciling.NetworkPolicyCreator) {
+		return "allow-dns", func(np *networkingv1.NetworkPolicy) (*networkingv1.NetworkPolicy, error) {
+			dnsPort := intstr.FromInt(53)
+			protoUdp := v1.ProtocolUDP
+			protoTcp := v1.ProtocolTCP
+
+			// dns access to node local dns cache
+			np.Spec = networkingv1.NetworkPolicySpec{
+				PolicyTypes: []networkingv1.PolicyType{
+					networkingv1.PolicyTypeIngress,
```
Comment (rastislavs) [inline pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: I am not getting why the ingress part of the policy is needed here?

**Comment by WeirdMachine** [comment]
It will definitely break something, but for dev this makes a lot of sense to figure out where we need to add or fix policies.

This should be enabled by default and I think it is okay to do so. It is relatively easy to add your own network policy. 

I agree with a setting in the cluster spec to to enable or disable network policies.

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go (hunk 6 above)
Comment (mate4st) [reply pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: Since the pod selector is empty it applies to all pods. Therefore it denies Ingress and Egress (Except of dns) to all Pods that do not have an explicit network policy applied. I will move this to another package as this is more a default policy for all pods than a dns specific one.

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+						},
+					},
+				},
+				// api access
+				Egress: []networkingv1.NetworkPolicyEgressRule{
+					{
+						To: []networkingv1.NetworkPolicyPeer{
+							{
+								IPBlock: &networkingv1.IPBlock{
+									CIDR: fmt.Sprintf("%s/32", k8sApiIP),
```
Comment (rastislavs) [inline pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go:116]: I think we should include `kubernetes` service `ClusterIP` (`10.240.16.1` with default settings) here as well, to have full compatibility with all CNIs:

https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors

> Cluster ingress and egress mechanisms often require rewriting the source or destination IP of packets. In cases where this happens, it is not defined whether this happens before or after NetworkPolicy processing, and the behavior may be different for different combinations of network plugin, cloud provider, Service implementation, etc.

> For egress, this means that connections from pods to Service IPs that get rewritten to cluster-external IPs may or may not be subject to ipBlock-based policies.

File: addons/openvpn/networkpolicy.yaml (hunk 5 above)
Comment (mate4st) [reply addons/openvpn/networkpolicy.yaml:1]: Yes, I was not sure if I can test it already as there is ongoing development. What do you think?

File: pkg/controller/user-cluster-controller-manager/resources/resources/usersshkeys/networkpolicy.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+package usersshkeys
+
+import (
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+)
+
+// NetworkPolicyCreator NetworkPolicy allows egress access to user ssh key agent.
```
Comment (rastislavs) [inline pkg/controller/user-cluster-controller-manager/resources/resources/usersshkeys/networkpolicy.go]: actually, it allows egress access **from** user ssh agent, right?

File: addons/csi/hetzner/hcloud-csi-1.6.0.yaml (hunk 3 above)
Comment (mate4st) [reply addons/csi/hetzner/hcloud-csi-1.6.0.yaml:390]: No, this does not allow egress traffic to the cluster. But let me double check. If I am wrong I will let you know

File: pkg/controller/user-cluster-controller-manager/resources/resources/usersshkeys/networkpolicy.go (hunk 8 above)
Comment (mate4st) [reply pkg/controller/user-cluster-controller-manager/resources/resources/usersshkeys/networkpolicy.go]: Yes let me rephrase it

File: addons/openvpn/networkpolicy.yaml (hunk 5 above)
Comment (rastislavs) [reply addons/openvpn/networkpolicy.yaml:1]: Some tweaks still needed, but it is already working in kubermatic dev environment.

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go (hunk 4 above)
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: I think this policy selects all the pods from all the namespaces and allows traffic only to DNS. 
Wouldn't that block traffic between the pods?

**Review by PratikDeoghare — COMMENTED** [review]
This should probably be behind a feature flag so that we can turn it off in case something goes wrong.

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go (hunk 4 above)
Comment (PratikDeoghare) [reply pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: https://kubernetes.io/docs/concepts/services-networking/network-policies/#isolated-and-non-isolated-pods 

> Pods become isolated by having a NetworkPolicy that selects them. Once there is any NetworkPolicy in a namespace selecting a particular pod, that pod will reject any connections that are not allowed by any NetworkPolicy. (Other pods in the namespace that are not selected by any NetworkPolicy will continue to accept all traffic.)

This is what it looks like on my cluster. 

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  creationTimestamp: "2021-12-02T19:01:39Z"
  generation: 1
  name: allow-dns
  namespace: kube-system
  resourceVersion: "128709"
  uid: e009eb81-5f2a-4524-a576-5e1b7423d162
spec:
  egress:
    - ports:
        - port: 53
          protocol: TCP
        - port: 53
          protocol: UDP
      to:
        - podSelector:
            matchLabels:
              app.kubernetes.io/name: kube-dns
    - ports:
        - port: 53
          protocol: TCP
        - port: 53
          protocol: UDP
      to:
        - ipBlock:
            cidr: 169.254.20.10/32
  podSelector: {}
  policyTypes:
    - Ingress
    - Egress
```

**Comment by PratikDeoghare** [comment]
@WeirdMachine 

Headsup: 

I have a cluster with 2 nodes. These pods for  `user-ssh-keys-agent, openvpn-client, envoy-agent`  are not coming up
for one of the nodes when I use the current branch. 

It could just be my hacks though.

**Comment by WeirdMachine** [comment]
> @WeirdMachine 
> 
> Headsup: 
> 
> I have a cluster with 2 nodes. These pods for  `user-ssh-keys-agent, openvpn-client, envoy-agent`  are not coming up
> for one of the nodes when I use the current branch. 
> 
> It could just be my hacks though. 
> 
> 

Do they throw any error?

**Comment by PratikDeoghare** [comment]
> 
Yes. Yes. OpenVPN health check fails. The other to fail to pull the image.

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go (hunk 7 above)
Comment (mate4st) [reply pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go:116]: Can this setting be changed? I did not found any usage of the ip address in the code, so i am not sure if i should start to hardcode it

Comment (rastislavs) [reply pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go:116]: It is the first address from the `cluster.Spec.ClusterNetwork.Services.CIDRBlocks[0]` range, which can be changed at cluster creation time. It appears as ClusterIP output of `kubectl get svc kubernetes` in the user cluster.

**Comment by embik** [comment]
A general question: Are we not scraping any of the locked down components via Prometheus for metrics? That would need to be in the ingress rules I think, so I'm curious.

File: addons/csi/hetzner/hcloud-csi-1.6.0.yaml (hunk 3 above)
Comment (PratikDeoghare) [reply addons/csi/hetzner/hcloud-csi-1.6.0.yaml:390]: > No, this does not allow egress traffic to the cluster

I don't follow. Which cluster are we talking about?

Comment (rastislavs) [reply addons/csi/hetzner/hcloud-csi-1.6.0.yaml:390]: So do we already know what is the difference between:
```
egress:
    - {}
```
and 
```
  egress:
    - to:
        - ipBlock:
            cidr: 0.0.0.0/0
```
?

File: pkg/controller/user-cluster-controller-manager/resources/reconciler.go [hunk 9]
```diff
@@ -857,6 +870,29 @@ func (r *reconciler) reconcileDeployments(ctx context.Context, data reconcileDat
...
+		kubesystem.DefaultNetworkPolicyCreator(),
+		coredns.KubeDNSNetworkPolicyCreator(data.clusterAddress.IP, int(data.clusterAddress.Port), data.k8sServiceApiIP.String()),
+	}
+
+	if r.userSSHKeyAgent {
+		namedNetworkPolicyCreatorGetters = append(namedNetworkPolicyCreatorGetters,
+			usersshkeys.NetworkPolicyCreator(data.clusterAddress.IP, int(data.clusterAddress.Port), data.k8sServiceApiIP.String()))
+	}
+
+	if r.isKonnectivityEnabled {
```
Comment (rastislavs) [inline pkg/controller/user-cluster-controller-manager/resources/reconciler.go:892]: haven't you forgotten about `konnectivity.NetworkPolicyCreator` here?

Comment (PratikDeoghare) [reply pkg/controller/user-cluster-controller-manager/resources/reconciler.go:892]: @WeirdMachine ^

File: pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8c.io/kubermatic/v2/pkg/resources"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+)
+
+// NetworkPolicyCreator NetworkPolicy allows all egress traffic
+func NetworkPolicyCreator() reconciling.NamedNetworkPolicyCreatorGetter {
+	return func() (string, reconciling.NetworkPolicyCreator) {
+		return "konnectivity", func(np *networkingv1.NetworkPolicy) (*networkingv1.NetworkPolicy, error) {
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go:30]: nit: a better name for this? maybe `konnectivity-allow-all`

File: pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package konnectivity
+
+import (
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go]: isn't this the other way round? :thinking: i.e. `k8s` imports are below `k8c` imports.

File: pkg/controller/user-cluster-controller-manager/resources/resources/node-local-dns/deamonset.go [hunk 12]
```diff
@@ -30,6 +30,8 @@ import (
 	"k8s.io/utils/pointer"
 )
 
+const CacheAddress = "169.254.20.10"
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/node-local-dns/deamonset.go]: This is called `NodeLocalDNSCacheAddress = "169.254.20.10"` in  pkg/resources/machinecontroller/deployment.go
Let us use the same name here too.

File: pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go (hunk 10 above)
Comment (mate4st) [reply pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go:30]: it does not allow all

File: pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go (hunk 4 above)
Comment (mate4st) [reply pkg/controller/user-cluster-controller-manager/resources/resources/core-dns/networkpolicy.go]: Yes it block traffic between the pods, thats the point of the network policy. The name was changed however to clarify that this is the default policy

File: pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go (hunk 10 above)
Comment (mate4st) [reply pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go:30]: It is blocking ingress and allow all egress

File: addons/csi/hetzner/hcloud-csi-1.6.0.yaml (hunk 3 above)
Comment (mate4st) [reply addons/csi/hetzner/hcloud-csi-1.6.0.yaml:390]: It is the same, unless you want to define ports, the networkpolicy editor is a bit confusing about it. What may makes sense here is to use cidr: 0.0.0.0/0 and use `except` and deny the cluster cidr. That way all egress to world would be allowed, but cluster access denied.

File: pkg/controller/user-cluster-controller-manager/resources/resources/node-local-dns/deamonset.go [hunk 13]
```diff
@@ -92,7 +91,7 @@ func DaemonSetCreator(registryWithOverwrite registry.WithOverwriteFunc) reconcil
 					ImagePullPolicy: corev1.PullAlways,
 					Args: []string{
 						"-localip",
-						CacheAddress,
+						machinecontroller.NodeLocalDNSCacheAddress,
```
Comment (PratikDeoghare) [inline pkg/controller/user-cluster-controller-manager/resources/resources/node-local-dns/deamonset.go]: I meant rename `CacheAddress` to `NodeLocalDNSCacheAddress`. 

Importing `"k8c.io/kubermatic/v2/pkg/resources/machinecontroller"` introduces dependency on the the machine controller.

File: pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go (hunk 10 above)
Comment (PratikDeoghare) [reply pkg/controller/user-cluster-controller-manager/resources/resources/konnectivity/networkpolicy.go:30]: how about `konnectivity-allow-egress` then? :-)

**Comment by kron4eg** [comment]
@WeirdMachine in release note it said "[ACTION REQUIRED]" (but missing this from the PR title), what action is required?

**Comment by WeirdMachine** [comment]
> @WeirdMachine in release note it said "[ACTION REQUIRED]" (but missing this from the PR title), what action is required?

Adding network policies for custom deployments in kube-system namespace maybe required.

**Comment by kron4eg** [comment]
> > @WeirdMachine in release note it said "[ACTION REQUIRED]" (but missing this from the PR title), what action is required?
> 
> Adding network policies for custom deployments in kube-system namespace maybe required.

Where to read about that, what policies exactly should be added?

**Comment by WeirdMachine** [comment]
I thought the release note is actually enough information. Anyway I added some more information to the description.

**Comment by WeirdMachine** [comment]
Updated description and removed action required as there is no effect on existing clusters.
